File: Tests/parallel.f90
      INTEGER FUNCTION test()
        IMPLICIT NONE
        INCLUDE "acc_testsuite.fh"
         INTEGER :: x, i_a, i_b, i_c, i_d, i_e, i_f, i_g, i_h, i_i, &
             i_j !Iterators
        INTEGER, PARAMETER :: PRIVATE_LOOPCOUNT = 1024
        REAL(8),DIMENSION(PRIVATE_LOOPCOUNT):: a, b, c !Data
        INTEGER :: errors
        errors = 0

        !Initilization
        CALL RANDOM_SEED()
        CALL RANDOM_NUMBER(a)
        CALL RANDOM_NUMBER(b)
        c = 0

        !$acc data copyin(a(1:PRIVATE_LOOPCOUNT), &
            !$acc b(1:PRIVATE_LOOPCOUNT)) &
            !$acc copy(c(1:PRIVATE_LOOPCOUNT))
          !$acc parallel
            !$acc loop
            DO i_a = 0, 1
              !$acc loop
              DO i_b = 0, 1
                !$acc loop
                DO i_c = 0, 1
                  !$acc loop
                  DO i_d = 0, 1
                    !$acc loop
                    DO i_e = 0, 1
                      !$acc loop
                      DO i_f = 0, 1
                        !$acc loop
                        DO i_g = 0, 1
                          !$acc loop
                          DO i_h = 0, 1
                            !$acc loop
                            DO i_i = 0, 1
                              !$acc loop
                              DO i_j = 0, 1
                                 c(i_a*512 + i_b*256 + i_c*128 + &
                                     i_d*64 + i_e*32 + i_f*16 + i_g*8 &
                                     + i_h*4 + i_i*2 + i_j + 1) = &
                                     a(i_a*512 + i_b*256 + i_c*128 + &
                                     i_d*64 + i_e*32 + i_f*16 + i_g*8 &
                                     + i_h*4 + i_i*2 + i_j + 1) + &
                                     b(i_a*512 + i_b*256 + i_c*128 + &
                                     i_d*64 + i_e*32 + i_f*16 + i_g*8 &
                                     + i_h*4 + i_i*2 + i_j + 1)
                              END DO
                            END DO
                          END DO
                        END DO
                      END DO
                    END DO
                  END DO
                END DO
              END DO
            END DO
          !$acc end parallel
        !$acc end data
        DO x = 1, PRIVATE_LOOPCOUNT
          IF (abs(c(x) - (a(x) + b(x))) .gt. PRECISION) THEN
            errors = errors + 1
          END IF
        END DO
        test = errors
      END


      PROGRAM test_kernels_async_main
      IMPLICIT NONE
      INTEGER :: failed, success !Number of failed/succeeded tests
      INTEGER :: num_tests,crosschecked, crossfailed, j
      INTEGER :: temp,temp1
      INCLUDE "acc_testsuite.fh"
      INTEGER test


      CHARACTER*50:: logfilename !Pointer to logfile
      INTEGER :: result

      num_tests = 0
      crosschecked = 0
      crossfailed = 0
      result = 1
      failed = 0

      !Open a new logfile or overwrite the existing one.
      logfilename = "OpenACC_testsuite.log"
!      WRITE (*,*) "Enter logFilename:"
!      READ  (*,*) logfilename

      OPEN (1, FILE = logfilename)

      WRITE (*,*) "######## OpenACC Validation Suite V 2.5 ######"
      WRITE (*,*) "## Repetitions:", N
      WRITE (*,*) "## Loop Count :", LOOPCOUNT
      WRITE (*,*) "##############################################"
      WRITE (*,*)

      WRITE (*,*) "--------------------------------------------------"
      WRITE (*,*) "Test of parallel"
      WRITE (*,*) "--------------------------------------------------"

      crossfailed=0
      result=1
      WRITE (1,*) "--------------------------------------------------"
      WRITE (1,*) "Test of parallel"
      WRITE (1,*) "--------------------------------------------------"
      WRITE (1,*)
      WRITE (1,*) "(Crosstests should fail)"
      WRITE (1,*)

      DO j = 1, N
        temp =  test()
        IF (temp .EQ. 0) THEN
          WRITE (1,*)  j, ". test successfull."
          success = success + 1
        ELSE
          WRITE (1,*) "Error: ",j, ". test failed."
          failed = failed + 1
        ENDIF
      END DO


      IF (failed .EQ. 0) THEN
        WRITE (1,*) "Directive worked without errors."
        WRITE (*,*) "Directive worked without errors."
        result = 0
        WRITE (*,*) "Result:",result
      ELSE
        WRITE (1,*) "Directive failed the test ", failed, " times."
        WRITE (*,*) "Directive failed the test ", failed, " times."
        result = failed * 100 / N
        WRITE (*,*) "Result:",result
      ENDIF
      CALL EXIT (result)
      END PROGRAM

File: Tests/parallel.cpp
#include "acc_testsuite.h"
#ifndef T1
//T1:parallel,V:1.0-2.7
int test1(){
    int err = 0;
    srand(SEED);
    real_t* a = new real_t[1024];
    real_t* b = new real_t[1024];
    real_t* c = new real_t[1024];

    for(int x = 0; x < 1024; ++x){
        a[x] = rand() / (real_t)(RAND_MAX / 10);
        b[x] = rand() / (real_t)(RAND_MAX / 10);
        c[x] = 0.0;
    }

    #pragma acc data copyin(a[0:1024], b[0:1024]) copy(c[0:1024])
    {
        #pragma acc parallel
        {
            #pragma acc loop
            for (int _0 = 0; _0 < 2; ++_0){
                #pragma acc loop
                for (int _1 = 0; _1 < 2; ++_1){
                    #pragma acc loop
                    for (int _2 = 0; _2 < 2; ++_2){
                        #pragma acc loop
                        for (int _3 = 0; _3 < 2; ++_3){
                            #pragma acc loop
                            for (int _4 = 0; _4 < 2; ++_4){
                                #pragma acc loop
                                for (int _5 = 0; _5 < 2; ++_5){
                                    #pragma acc loop
                                    for (int _6 = 0; _6 < 2; ++_6){
                                        #pragma acc loop
                                        for (int _7 = 0; _7 < 2; ++_7){
                                            #pragma acc loop
                                            for (int _8 = 0; _8 < 2; ++_8){
                                                #pragma acc loop
                                                for (int _9 = 0; _9 < 2; ++_9){
                                                    c[_0*512 + _1*256 + _2*128 + _3*64 + _4*32 + _5*16 + _6*8+ _7*4 + _8*2 + _9] =
                                                    a[_0*512 + _1*256 + _2*128 + _3*64 + _4*32 + _5*16 + _6*8+ _7*4 + _8*2 + _9] +
                                                    b[_0*512 + _1*256 + _2*128 + _3*64 + _4*32 + _5*16 + _6*8+ _7*4 + _8*2 + _9];
                                                }
                                            }
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
    }

    for (int x = 0; x < 1024; ++x){
        if(fabs(c[x] - (a[x] +b[x])) > PRECISION){
            err = 1;
        }
    }

    return err;
}
#endif

int main(){
    int failcode = 0;
    int failed;
#ifndef T1
    failed = 0;
    for (int x = 0; x < NUM_TEST_CALLS; ++x){
        failed = failed + test1();
    }
    if (failed != 0){
        failcode = failcode + (1 << 0);
    }
#endif
    return failcode;
}
